### PR TITLE
Add stats to cache and watch access

### PR DESCRIPTION
**Pull request by jcantrill**
This PR adds stat info regarding cache hits to superced: #87 

cc @portante

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 source 'https://rubygems.org'
 
-gem 'codeclimate-test-reporter', :group => :test, :require => nil
+gem 'codeclimate-test-reporter', '=0.6.0', :group => :test, :require => nil
```
Comment (richm) [inline Gemfile]: why not just '< 1.0.0'

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 2]
```diff
@@ -16,6 +16,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+require File.dirname(__FILE__) + '/kubernetes_metadata_stats'
```
Comment (richm) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: Use http://ruby-doc.org/core-2.0.0/Kernel.html#method-i-require_relative instead?

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 3]
```diff
@@ -110,9 +112,52 @@ def parse_namespace_metadata(namespace_object)
     def get_pod_metadata(namespace_name, pod_name)
       begin
         metadata = @client.get_pod(pod_name, namespace_name)
-        return if !metadata
-        return parse_pod_metadata(metadata)
-      rescue KubeException
+        if !metadata
```
Comment (richm) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: `unless` instead of `if !`?

Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: or you could combine lines 114 and 115:
```ruby
unless (metadata = @client.get_pod(pod_name, namespace_name))
```

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 4]
```diff
@@ -110,9 +112,52 @@ def parse_namespace_metadata(namespace_object)
...
+      @cur_time = Time.now
+      return if @curr_time - @prev_time < @stats_interval
+      @prev_time = @curr_time
+      log.info(@stats)
+    end
+
+    def get_namespace_metadata(namespace_name)
+      begin
+        metadata = @client.get_namespace(namespace_name)
+        if !metadata
```
Comment (richm) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]:
```ruby
unless (metadata = @client.get_namespace(namespace_name))
```

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 5]
```diff
@@ -110,15 +112,61 @@ def parse_namespace_metadata(namespace_object)
...
+        log.debug "Exception encountered fetching pod metadata from Kubernetes API #{@apiVersion} endpoint #{@kubernetes_url}: #{e.message}"
+        nil
+      end
+    end
+
+    def dump_stats
+      @curr_time = Time.now
+      return if (@curr_time - @prev_time).to_i < @stats_interval
+      @prev_time = @curr_time
+      log.info(@stats)
```
Comment (portante) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: Can we also dump info about the caches themselves?  How many entries currently maybe?

File: lib/fluent/plugin/kubernetes_metadata_stats.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+#
+# Fluentd Kubernetes Metadata Filter Plugin - Enrich Fluentd events with
+# Kubernetes metadata
+#
+# Copyright 2015 Red Hat, Inc.
```
Comment (portante) [inline lib/fluent/plugin/kubernetes_metadata_stats.rb]: Copyright 2017?

File: test/plugin/test_cache_stats.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+#
+# Fluentd Kubernetes Metadata Filter Plugin - Enrich Fluentd events with
+# Kubernetes metadata
+#
+# Copyright 2015 Red Hat, Inc.
```
Comment (portante) [inline test/plugin/test_cache_stats.rb]: Copyright 2017?

File: lib/fluent/plugin/kubernetes_metadata_stats.rb (hunk 6 above)
Comment (jcantrill) [reply lib/fluent/plugin/kubernetes_metadata_stats.rb]: Most likely can not update this as the other lint checks usually fail when I have tried to update the year

File: test/plugin/test_cache_stats.rb (hunk 7 above)
Comment (jcantrill) [reply test/plugin/test_cache_stats.rb]: Most likely can not update this as the other lint checks usually fail when I have tried to update the year

File: lib/fluent/plugin/filter_kubernetes_metadata.rb (hunk 5 above)
Comment (jcantrill) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: Not sure what the ask is here.  By dumping the stats you will get a list of all the entries and the associated counts.  What additional value is provided by the raw number you are requesting?

Comment (portante) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: I was thinking we should emit the size of each cache at the least, to know if we are actually evicting on new entries or only on ttl.

**Pull request merged**